### PR TITLE
Add fake_quantize_op.

### PR DESCRIPTION
**Pull request by achao2013**
add quant code for test

### CODE REVIEW

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    // if (ctx->HasInput("InMovingScale")) {
+    ctx->SetOutputDim("OutMovingScale", ctx->GetInputDim("InMovingScale"));
+    //}
+    // if (ctx->HasInput("InScales")) {
+    PADDLE_ENFORCE(ctx->HasOutput("OutScales"), "");
+    ctx->SetOutputDim("OutScales", ctx->GetInputDim("InScales"));
+    // PADDLE_ENFORCE_EQ(ctx->Inputs("InScales")[0],
+    // ctx->Outputs("OutScales")[0],
+    //                  "Mean and MeanOut should share the same memory");
+    //}
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cc:46]: Please remove the commented lines.

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                 const framework::VariableNameMap &outputs,
+                 const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"),
+                   "Input(X) of FakeQuantizeOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"),
+                   "Output(Out) of FakeQuantizeOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("OutMovingScale"), "");
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cc]: Please add the error message.

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    // if (ctx->HasInput("InMovingScale")) {
+    ctx->SetOutputDim("OutMovingScale", ctx->GetInputDim("InMovingScale"));
+    //}
+    // if (ctx->HasInput("InScales")) {
+    PADDLE_ENFORCE(ctx->HasOutput("OutScales"), "");
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cc]: Please add the error message.

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    ctx->SetOutputDim("Out", ctx->GetInputDim("X"));
+    ctx->ShareLoD("X", /*->*/ "Out");
+  }
+};
+
+class FakeQuantizeOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "(Tensor) Input tensor of scale operator.");
+    AddInput("InScales", "(Tensor) scale buffer").AsDispensable();
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cc]: Please add more comments for why this argument is optional. When need it and when don't need it.  The same is the following.

File: paddle/fluid/operators/fake_quantize_op.cu [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+#include <string>
+#include "paddle/fluid/operators/fake_quantize_op.h"
+#include "paddle/fluid/platform/cuda_primitives.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+__global__ void find_abs_max_kernel(const int n, const T* in, T* out) {
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cu]: find_abs_max_kernel -> FindAbsMaxKernel

Please follow Google C++ code style: https://google.github.io/styleguide/cppguide.html#Function_Names

Please modify other code with the same problem.

File: paddle/fluid/operators/fake_quantize_op.cu [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  }
+  if (tid == 0) {
+    out[blockIdx.x] = shared_max_data[0];
+  }
+}
+
+float find_abs_max_gpu(const platform::CUDADeviceContext& ctx,
+                       const float* array, int length) {
+  float host_max;
+  int NUM_THREADS = 1024;
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cu]: NUM_THREADS -> kNumTheads Please follow Goolge code style.

File: paddle/fluid/operators/fake_quantize_op.cu [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+                               NUM_THREADS * sizeof(float), ctx.stream()>>>(
+      length, array, device_max);
+  find_abs_max_kernel<
+      float><<<1, NUM_THREADS, NUM_THREADS * sizeof(float), ctx.stream()>>>(
+      gridDimx, device_max, device_max);
+  PADDLE_ENFORCE_EQ(
+      cudaMemcpy(&host_max, device_max, sizeof(float), cudaMemcpyDeviceToHost),
+      cudaSuccess, "cudaMemcpy failed");
+  return host_max;
+}
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cu:74]: Maybe can use thrust::reduce + thrust::max_element to find the maximum value for more simply.

File: paddle/fluid/operators/fake_quantize_op.h [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+      auto* oci_tensor = context.Output<framework::Tensor>("OutCurrentIter");
+      oci_tensor->mutable_data<T>(
+          context.Input<framework::Tensor>("InCurrentIter")->place());
+    }
+
+    T scale = T(1);
+    int window_size = context.Attr<int>("window_size");
+    int bit_length = context.Attr<int>("bit_length");
+    int bin_cnt = std::pow(2, bit_length - 1) - 1;
+    LOG(ERROR) << "bin_cnt:" << bin_cnt;
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.h]: Remove this line.

File: paddle/fluid/operators/fake_quantize_op.h [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        scale = moving_scale->data<T>()[0];
+      } else {
+        auto* it = context.Input<framework::Tensor>("InCurrentIter");
+        auto* iter = context.Output<framework::Tensor>("OutCurrentIter");
+        const int* last_iter = it->data<int>();
+        int* current_iter = iter->mutable_data<int>(platform::CPUPlace());
+        auto* scale_list = context.Output<framework::Tensor>("OutScales");
+        auto* saving_scale =
+            context.Output<framework::Tensor>("OutMovingScale");
+        scale = find_abs_max(const_cast<framework::Tensor*>(in), in->numel());
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.h]: Here can use Eigen's method：

https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/adamax_op.h#L57

File: paddle/fluid/operators/fake_quantize_op.h [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+      } else {
+        scale = find_abs_max(const_cast<framework::Tensor*>(in), in->numel());
+        auto* saving_scale =
+            context.Output<framework::Tensor>("OutMovingScale");
+        scale = find_moving_average_abs_max(
+            const_cast<framework::Tensor*>(moving_scale), saving_scale, scale);
+      }
+    }
+
+    apply_saturate(const_cast<framework::Tensor*>(in), tensor, -scale, scale);
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.h]: Also can refer https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/clip_op.h#L70  for more simply.

**Review by qingqing01 — COMMENTED** [review]
Need unit testing.

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    AddInput("InMovingScale", "last scale").AsDispensable();
+    AddInput("InCurrentIter", "last Iter").AsDispensable();
+    AddOutput("Out", "(Tensor) Output tensor of scale operator.");
+    AddOutput("OutScales", "(Tensor) scale buffer").AsDispensable();
+    AddOutput("OutMovingScale", " current scale");
+    AddOutput("OutCurrentIter", "current Iter").AsDispensable();
+    AddComment(R"DOC(
+FakeQuantize operator
+
+$$Out = scale*X$$
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cc]: Need comments for how to calculate `scale`.

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+)DOC");
+    AddAttr<std::string>("quantize_type",
```
Comment (qingqing01) [inline paddle/fluid/operators/fake_quantize_op.cc]: quantize_type -> scale_type for more accurate ?

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 1 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cc:46]: the comment is for test of python , the commented lines is used for train

File: paddle/fluid/operators/fake_quantize_op.cu (hunk 7 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cu:74]: this will be slow

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 12 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cc]: if the quantization method is non-uniform, scale is not  need, so i think this should  not be scale_type

File: paddle/fluid/operators/fake_quantize_op.h (hunk 9 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.h]: this cwiseMax is an elemwise max operation, i need a reduce max op .

Comment (qingqing01) [reply paddle/fluid/operators/fake_quantize_op.h]: Please refer https://github.com/RobotLocomotion/eigen-mirror/tree/master/unsupported/Eigen/CXX11/src/Tensor#-reduceconst-dimensions-new_dims-const-reducer-reducer

or 

 https://github.com/RobotLocomotion/eigen-mirror/tree/master/unsupported/Eigen/CXX11/src/Tensor#-maximum

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 12 above)
Comment (qingqing01) [reply paddle/fluid/operators/fake_quantize_op.cc]: 我理解`quantize_type`一般指： Abs-Max，或者Min-Max等不同的量化方式。
而这里，这个attr是想来标示，计算scale的方式吧？

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 2 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cc]: done

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 3 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cc]: done

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 4 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cc]: done

File: paddle/fluid/operators/fake_quantize_op.cu (hunk 5 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cu]: done

File: paddle/fluid/operators/fake_quantize_op.cu (hunk 6 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cu]: done

File: paddle/fluid/operators/fake_quantize_op.h (hunk 8 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.h]: done

File: paddle/fluid/operators/fake_quantize_op.h (hunk 9 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.h]: done

File: paddle/fluid/operators/fake_quantize_op.h (hunk 10 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.h]: done

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 11 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cc]: done

File: paddle/fluid/operators/fake_quantize_op.cc (hunk 12 above)
Comment (achao2013) [reply paddle/fluid/operators/fake_quantize_op.cc]: 如果是非均匀量化， 那浮点输入和定点输出可能是个函数，或者离散数值映射，就没有scale操作了

**Comment by qingqing01** [comment]
@achao2013 

> [07:05:06]W:	 [Step 1/1] /paddle/paddle/fluid/operators/distributed/rpc_client.h:61:26: error: 'FLAGS_grpc_deadline' was not declared in this scope
[07:05:06]W:	 [Step 1/1]        int64_t time_out = FLAGS_grpc_deadline) = 0;

CI没有通过，需要更新至最新develop代码。

**Comment by qingqing01** [comment]
```
[07:00:33]	294/359 Test #290: test_fake_quantize_op ................................***Failed    5.52 sec
[07:00:33]	test_fake_quantize_op failed
[07:00:33]	E
[07:00:33]	======================================================================
[07:00:33]	ERROR: test_check_output (test_fake_quantize_op.TestFakeQuantizeOp)
[07:00:33]	----------------------------------------------------------------------
[07:00:33]	Traceback (most recent call last):
[07:00:33]	  File "/paddle/build/python/paddle/fluid/tests/unittests/test_fake_quantize_op.py", line 47, in test_check_output
[07:00:33]	    self.check_output()
[07:00:33]	  File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 325, in check_output
[07:00:33]	    self.check_output_with_place(place, atol)
[07:00:33]	  File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 263, in check_output_with_place
[07:00:33]	    outs, fetch_list = self._calc_output(place)
[07:00:33]	  File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 226, in _calc_output
[07:00:33]	    inputs = self._get_inputs(block)
[07:00:33]	  File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 211, in _get_inputs
[07:00:33]	    return self._get_io_vars(block, self.inputs)
[07:00:33]	  File "/paddle/build/python/paddle/fluid/tests/unittests/op_test.py", line 207, in _get_io_vars
[07:00:33]	    inputs[name] = block.var(name)
[07:00:33]	  File "/paddle/build/python/paddle/fluid/framework.py", line 900, in var
[07:00:33]	    raise ValueError("var %s not in this block" % name)
[07:00:33]	ValueError: var InCurrentScale not in this block
[07:00:33]	
[07:00:33]	----------------------------------------------------------------------
[07:00:33]	Ran 1 test in 0.002s
[07:00:33]	
[07:00:33]	FAILED (errors=1)
```

The unit testing did not pass.

**Review by qingqing01 — APPROVED** [review]
Approved. @dangqingqing will refine and add more unit testing.

**Pull request merged**